core/src/main/java/ch/cyberduck/core/ProtocolFactory.java

```diff
@@ -220,24 +221,50 @@ public Protocol forName(final String identifier, final String provider) {
-                                // Matching vendor string usage in CLI
-                                enabled.stream().sorted(new DeprecatedProtocolComparator()).filter(protocol -> StringUtils.equals(protocol.getProvider(), identifier)).findFirst().orElse(
-                                        // Fallback for bug in 6.1
-                                        enabled.stream().sorted(new DeprecatedProtocolComparator()).filter(protocol -> StringUtils.equals(String.format("%s-%s", protocol.getIdentifier(), protocol.getProvider()), identifier)).findFirst().orElse(
-                                                // Matching bundled first with identifier match
-                                                enabled.stream().sorted(new DeprecatedProtocolComparator()).filter(protocol -> protocol.isBundled() && StringUtils.equals(protocol.getIdentifier(), identifier)).findFirst().orElse(
-                                                        // Matching scheme with fallback to generic protocol type
-                                                        this.forScheme(enabled, identifier, enabled.stream().sorted(new DeprecatedProtocolComparator()).filter(protocol -> StringUtils.equals(protocol.getType().name(), identifier)).findFirst().orElse(null))
-                                                )
-                                        )
-                                )
-                        )
-                );
+                return protocol;
+            }
+            /*
+            All following cases are required for fallback/broader searches.
+            There may be other, better matches in the stream
+            Each state represents findFirst stream semantic, and overwrites the result of lower matches
+             */
+            else if(state < 4 && StringUtils.equals(protocol.getIdentifier(), identifier)
+                    && StringUtils.equals(protocol.getProvider(), provider)) {
+                // Matching vendor string for third party profiles
+                match = protocol;
+                state = 4;
+            }
+            else if(state < 3 && StringUtils.equals(protocol.getProvider(), identifier)) {
+                // Matching vendor string usage in CLI
+                match = protocol;
+                state = 3;
+            }
+            else if(state < 2 && StringUtils.equals(
+                    String.format("%s-%s", protocol.getIdentifier(), protocol.getProvider()), identifier)) {
+                // Fallback for bug in 6.1
+                match = protocol;
+                state = 2;
+            }
+            else if(state < 1 && protocol.isBundled() && StringUtils.equals(protocol.getIdentifier(), identifier)) {
+                // Matching bundled first with identifier match
+                match = protocol;
+                state = 1;
+            }
+            else if(null == match && StringUtils.equals(protocol.getType().name(), identifier)) {
```

Comment:
Necessary? `state` remains `0` and thus overridden in line 266 if there is no other match.

Reply:
It is, because the state=0 check should fire only if nothing in the loop ever matched.
Fixed in #726bfa8

Reply:
Add test coverage?

Reply:
Specifically:
The `null == match`-check in this loop is this part of the previous implementation:
```
enabled.stream().sorted(new DeprecatedProtocolComparator()).filter(protocol -> StringUtils.equals(protocol.getType().name(), identifier)).findFirst().orElse(null)
```
which is used later in
```
this.forScheme(enabled, identifier, …)
```
As the forScheme-fallback is only used after all previous streams did not return a match, this must be fetched later, and overwrite the type-name match.